src/Concerns/Exportable.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace PodPoint\MailExport\Concerns;
+
+use Illuminate\Contracts\Mail\Mailer as MailerContract;
+use Illuminate\Mail\Mailable;
+use PodPoint\MailExport\Contracts\ShouldExport;
+
+/**
+ * @mixin Mailable
+ */
+trait Exportable
+{
+    /**
+     * Send the message using the given mailer.
+     *
+     * @param  \Illuminate\Contracts\Mail\Mailer  $mailer
+     * @return void
+     */
+    public function send(MailerContract $mailer)
+    {
+        $this->withSwiftMessage(function ($message) {
+            $message->_shouldStore = $this instanceof ShouldExport;
+            $message->_storageDisk = $this->storageDisk();
+            $message->_storagePath = $this->storagePath();
+            $message->_storageFilename = $this->storageFilename();
```

Comment:
I've tried a lot of things but this was the easiest and wouldn't disrupt anything as this dynamic properties would die at the end of the runtime.

I've tried a lot of things but it involved a lot of pain around extending the Laravel `Mailer` and `Message` classes and  because it changed so much during Laravel's lifetime it would make it very hard to build a package.

I'll try another approach with Swiftmailer's headers.

Reply:
Okay using a custom header with Swiftmailer works but it obviously leaks into the actual email sent out.

```
Message-ID: <559b47f2a6437ae80ed49a723e08517b@swift.generated>
Date: Mon, 29 Mar 2021 20:02:55 +0000
Subject: This is the subject
From: jane@example.com
To: Jane Doe <jane@example.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable
X-Mail-Export: storage; disk=local; path=some_path; filename=some_filename

This is the body.
```

Notice the `X-Mail-Export` header.

I don't think there is really a simpler/quicker way to do it than piggy-backing on the Swift Message unfortunately.

I'll keep looking if there is something we can use offered by Swiftmailer that doesn't leak into the actual message.

Reply:
This is another approach using the `Cache` to temporarily store the storage options:

https://github.com/Pod-Point/laravel-mail-export/compare/feature/SWP-2200-lifecycle-cache?expand=1

Reply:
Thank you for exhausting other options - agree it's not ideal exposing configuration options into the actual mail. I don't think it's the end of the world if we use property overriding, it's certainly simpler - it just feels a little off - but I guess it's just taking advantage of the dynamic language so 🤷🏻

Reply:
Okay so that eliminates piggy backing on Swiftmailer headers.

What about between dynamic properties and the Cache approach? How do you feel?

Reply:
Between the two, I'd say dynamic properties, as that doesn't introduce any further reliance on other system services and doesn't require users to go through updating their mailables to include `export()`.

Reply:
I think that answers my most recent question 😅

Reply:
Wait I think I got confused between data object and value object. I think I meant value object.

Reply:
Yeah me too - I thought it was value object, but then the spatie library that I remembered as bringing it to my attention was originally [value-object](https://packagist.org/packages/spatie/value-object) but then became [data-transfer-object](https://packagist.org/packages/spatie/data-transfer-object)...

Reply:
Looks like we were both right `data-value-transfer-data-object-value` 😅